### ISSUE TITLE
Feature Request: Web Key Directory support

### ISSUE THREAD

**Issue by ManoftheSea** [open-keychain/open-keychain#2560]
https://wiki.gnupg.org/WKD

1. The sender's mail client checks a "well known" URL on the domain of the recipient.
2. If a public key is available for that mail address, will be downloaded via HTTPS.
3. The downloaded pubkey can now be used without further user interaction.

## Current Behavior
Currently, only key servers are supported, or trying to copy files over to devices that more and more don't have SD cards.

## Possible Solution
Add a UI element asking to search WKD for a user email.  Get the Domain from the entered string.  Look up the openpgpkey.<domain> A record OR the _openpgp._tcp.<domain> SRV record.  Use the result to find the .well-known/openpgpkey/hu/key-hash file for the user.  Import that public key, offering the user the ability to assign it some trust.

## Context
As I update my own key, I publish it to the Web Key Service for my domain, much like publishing to a keyserver.  This puts the trust in the key found by the servers somewhere between TOFU and in-person verification.

**Comment by d-g** [comment]
Derek <notifications@github.com> wrote:
> Look up the openpgpkey.<domain> A record OR the _openpgp._tcp.<domain> SRV record.

Interesting.  Where does the latter scheme come from?

> Use the result to find the .well-known/openpgpkey/hu/key-hash file

Nope, this is a path for just ‘example.org’ without ‘openpgpkey’ (which should be supported well).  For ‘openpgpkey.example.org’ it would be ‘/.well-known/openpgpkey/example.org/hu/${username-hash}’.


I believe, this is a good illustration to the major obstacle to adoption of WKD (besides the fact, that itʼs less useful than keyservers, and is inferior trustwise both to keyservers and to DNS-based solutions, known as ‘pka’, ‘cert’ and ‘dane’ in GPG) — there is still no standard published.

**Comment by Valodim** [comment]
This is already supported: when you search for an email address, WKD will also be checked for a key.

This can be enabled and disabled in the search settings, next to keyservers and keybase.io.

**Comment by ManoftheSea** [comment]
@d-g I apologize, the SRV resource records come from versions 3-6 of https://tools.ietf.org/html/draft-koch-openpgp-webkey-service-06, but removed in version 7, as the "advanced" DNS lookup is added.  Which is as you describe.  I looked for this feature within the app and did not recognize it, but I am happy to see it already exists and is useful.
At https://wiki.gnupg.org/AutomatedEncryption#Level_2, it describes that WKD is more trustworthy than simple TOFU, and I understand it that way as well.  If I tell someone to "send me a message at myemail@example.com", they may have slightly more trust that when the domain provides them a key to send to me, it is my key, than if they simply found it (and 3 others) on some random keyserver.
@Valodim  Thank you very much.  Perhaps the fault continues to be my own, but I don't think the feature is complete.  How does one update a key via WKD?  I would think I could open an existing key and interact with a user id to request a WKD lookup.  Once I have a contact with the user id, doing a search does not appear to check the WKD source.  Could you please advise?

**Comment by d-g** [comment]
Derek <notifications@github.com> wrote:
> At https://wiki.gnupg.org/AutomatedEncryption#Level_2, it describes that WKD is more trustworthy than simple TOFU

Of course, it is.  _Anything_ is more trustworthy than simple TOFU.

> If I tell someone to "send me a message at myemail@example.com", they may have slightly more trust that when the domain provides them a key to send to me, it is my key

Sure, people indeed do that, and I hope you realise, that it is a _problem_.

Itʼs not a rule of course, but in real life your domain provider strongly tends to be your mail provider as well, i. e. the only party, besides a mail provider of your correspondent, you are protecting against.

In other words, in most real use-cases WKD adds nothing in terms of trust on top of just mailing a key (by rolling it into signature¹, by Autocrypt or informally).  It adds convenience, of course.

> than if they simply found it on some random keyserver.

On the contrary, a random keyserver tends to be not controlled by your mail provider.

-
¹ ‘include-key-block’ option in GPG.
